database/engine/rrdengineapi.c

```diff
@@ -74,10 +74,14 @@ void rrdeng_metric_init(RRDDIM *rd, uuid_t *dim_uuid)
         if (unlikely(find_or_generate_guid(rd, rd->state->metric_uuid, GUID_TYPE_DIMENSION,
                                            replace_instead_of_generate))) {
             errno = 0;
-            error("FAILED to generate GUID for %s", rd->id);
-            freez(rd->state->metric_uuid);
-            rd->state->metric_uuid = NULL;
-            fatal_assert(0);
+            error("FAILED to reuse GUID for %s", rd->id);
+            if (unlikely(find_or_generate_guid(rd, rd->state->metric_uuid, GUID_TYPE_DIMENSION, 0))) {
+                errno = 0;
+                error("FAILED to generate GUID for %s", rd->id);
+                freez(rd->state->metric_uuid);
+                rd->state->metric_uuid = NULL;
+                fatal_assert(0);
```

Comment:
Why kill the agent?

Reply:
This should be an impossible state. You can't have a dbengine dimension without a UUID, it has to fail here or it will fail later. The collectors expect to be able to use dimensions, there is no other failure mode here unfortunately.

Reply:
Just because it should be impossible is not a strong enough reason. If it was actually impossible then it would be valid to ignore it. If we have code to trap it thenit just seems to be unlikely, which is not the same. If something goes wrong - what is the least harmful way of dealing with it?

Reply:
Why not mark the dimension as disabled / in error - and allow the agent to continue?

Reply:
When I detect state corruption I prefer to fail fast, before propagating the corruption to other components. This way it's easier to debug and you can do less damage to user data. That's my personal preference though.

Reply:
In some systems that is desirable, especially if they are short-lived and interactive (e.g. command line tools). In others it can be quite harmful. We should aim for graceful degradation of service, and best-effort delivery where possible. In this specific case - what can cause the generation attempt to fail - and what does it tell us about the state of the system? i.e. does it make sense to continue, does it make sense to abort.

Reply:
In this specific case  (i.e. after the collision has been detected) it will only fail if the newly generated GUID already exists in the map i.e. we have managed to generate a duplicate GUID.

In case of memory allocation errors with mallocz or JudyHSIns error,  it will never return because fatal() is called.

Reply:
That seems insanely unlikely. Why not just repeat it if that happens instead of aborting the agent?